packages/rspack/src/lib/EnvironmentPlugin.ts

```diff
@@ -1,50 +1,52 @@
-/*
-	MIT License http://www.opensource.org/licenses/mit-license.php
-	Authors Simen Brekken @simenbrekken, Einar Löve @einarlove
-*/
-
-"use strict";
+/**
+ * The following code is modified based on
+ * https://github.com/webpack/webpack/blob/4b4ca3b/lib/EnvironmentPlugin.js
+ *
+ * MIT Licensed
+ * Author Tobias Koppers @sokra
+ * Copyright (c) JS Foundation and other contributors
+ * https://github.com/webpack/webpack/blob/main/LICENSE
+ */
 
 import { DefinePlugin } from "../builtin-plugin";
+import WebpackError from "./WebpackError";
 
-const WebpackError = require("./WebpackError");
+import type { Compiler } from "../Compiler";
 
-// /** @typedef {import("./Compiler")} Compiler */
-// /** @typedef {import("./DefinePlugin").CodeValue} CodeValue */
-/** @typedef {any} Compiler */
-/** @typedef {any} CodeValue */
+// Waiting to adapt > import("./DefinePlugin").CodeValue
+type CodeValue = any;
 
 class EnvironmentPlugin {
-	// @ts-expect-error
-	constructor(...keys) {
+	keys: string[];
+	defaultValues: Record<string, string>;
+
+	constructor(...keys: string[] | [Record<string, string> | string]) {
 		if (keys.length === 1 && Array.isArray(keys[0])) {
 			this.keys = keys[0];
 			this.defaultValues = {};
-		} else if (keys.length === 1 && keys[0] && typeof keys[0] === "object") {
```

Comment:
`[null].length === 1`, so... this modification might not be the same as before
just a corner case